npm@5.3: npm prune --production fails

#### I'm opening this issue because:

  - [ ] npm is producing an incorrect install.
  - [ ] npm is crashing.
  - [x] npm is doing something I don't understand.
  - [ ] Other (_see below for feature requests_):

#### What's going wrong?
`npm prune --production` fails with `npm ERR! May not delete: /path/node_modules/.bin`
`NODE_ENV=production npm prune` fails with `npm ERR! May not delete: /path/node_modules/.bin`

#### How can the CLI team reproduce the problem?
Run `NODE_ENV=production npm prune` or  `npm prune --production` in any project which has a `node_modules/.bin` folder with npm@5.3.0

npm@5.2.0 is working as expected.

### supporting information:

 - `npm -v` prints: 5.3.0
 - `node -v` prints: v8.1.4
 - `npm config get registry` prints: https://registry.npmjs.org/
 - Windows, OS X/macOS, or Linux?: Linux 3.13.0-123-generic

Super easy reproduction steps ✨ 

```bash
mkdir prune-test && cd prune-test
npm init -y
npm install -D yes
npm prune --production
```

Output:

```
 prune-test ツ npm prune --production
npm WARN prune-test@1.0.0 No description
npm WARN prune-test@1.0.0 No repository field.

npm ERR! May not delete: C:\msys64\home\amwenger\Dev\prune-test\node_modules\.bin

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\amwenger\AppData\Roaming\npm-cache\_logs\2017-07-25T10_20_15_408Z-debug.log
```

--- a/lib/unbuild.js
+++ b/lib/unbuild.js
@@ -60,7 +60,9 @@ function rmStuff (pkg, folder, cb) {
   // otherwise, then bins are in folder/../.bin
   var parent = pkg.name[0] === '@' ? path.dirname(path.dirname(folder)) : path.dirname(folder)
   var gnm = npm.dir
-  var top = gnm === parent
+  // gnm might be an absolute path, parent might be relative
+  // this checks they're the same directory regardless
+  var top = path.relative(gnm, parent) === ''
 
   log.verbose('unbuild rmStuff', pkg._id, 'from', gnm)
   if (!top) log.verbose('unbuild rmStuff', 'in', parent)

--- a/test/tap/prune-dev-dep-with-bins.js
+++ b/test/tap/prune-dev-dep-with-bins.js
@@ -1,0 +1,105 @@
+'use strict'
+var fs = require('fs')
+var path = require('path')
+var test = require('tap').test
+var Tacks = require('tacks')
+var File = Tacks.File
+var Dir = Tacks.Dir
+var common = require('../common-tap.js')
+var testdir = path.join(__dirname, path.basename(__filename, '.js'))
+
+var fixture = new Tacks(
+  Dir({
+    node_modules: Dir({
+      'yes': Dir({
+        'package.json': File({
+          _requested: {
+            rawSpec: 'file:///mods/yes'
+          },
+          dependencies: {},
+          bin: {
+            'yes': 'yes.js'
+          },
+          name: 'yes',
+          version: '1.0.0'
+        }),
+        'yes.js': File('while (true) { console.log("y") }')
+      }),
+      '.bin': Dir({
+        // verbose, but needed for `read-cmd-shim` to properly identify which
+        // package this belongs to
+        'yes': File(
+          '#!/bin/sh\n' +
+          'basedir=$(dirname "$(echo "$0" | sed -e \'s,\\\\,/,g\')")\n' +
+          '\n' +
+          'case `uname` in\n' +
+          '    *CYGWIN*) basedir=`cygpath -w "$basedir"`;;\n' +
+          'esac\n' +
+          '\n' +
+          'if [ -x "$basedir/node" ]; then\n' +
+          '  "$basedir/node"  "$basedir/../yes/yes.js" "$@"\n' +
+          '  ret=$?\n' +
+          'else\n' +
+          '  node  "$basedir/../yes/yes.js" "$@"\n' +
+          '  ret=$?\n' +
+          'fi\n' +
+          'exit $ret\n'),
+        'yes.cmd': File(
+          '@IF EXIST "%~dp0\node.exe" (\n' +
+          '"%~dp0\\node.exe"  "%~dp0\\..\\yes\\yes.js" %*\n' +
+          ') ELSE (\n' +
+          '@SETLOCAL\n' +
+          '@SET PATHEXT=%PATHEXT:;.JS;=;%\n' +
+          'node  "%~dp0\..\yes\yes.js" %*')
+      })
+    }),
+    'package.json': File({
+      name: 'test',
+      version: '1.0.0',
+      devDependencies: {
+        'yes': 'file:///mods/yes'
+      }
+    })
+  })
+)
+
+function setup () {
+  cleanup()
+  fixture.create(testdir)
+}
+
+function cleanup () {
+  fixture.remove(testdir)
+}
+
+test('setup', function (t) {
+  setup()
+  t.end()
+})
+
+function readdir (dir) {
+  try {
+    return fs.readdirSync(dir)
+  } catch (ex) {
+    if (ex.code === 'ENOENT') return []
+    throw ex
+  }
+}
+
+test('prune cycle in dev deps', function (t) {
+  common.npm(['prune', '--production', '--json'], {cwd: testdir}, function (err, code, stdout, stderr) {
+    if (err) throw err
+    t.is(code, 0, 'prune finished successfully')
+    t.like(JSON.parse(stdout), {removed: [{name: 'yes'}]}, 'removed the right modules')
+    var dirs = readdir(testdir + '/node_modules').sort()
+    // bindirs are never removed, it's ok for them to remain after prune
+    t.same(dirs, ['.bin'])
+    t.end()
+  })
+})
+
+test('cleanup', function (t) {
+  cleanup()
+  t.end()
+})
+

